File: euler/benchmarks/problem12/nim/main.nim
import tables, sets, sequtils

type
  PrimeGeneratorState = ref object
    n: int64
    last: int64
    sieve: TableRef[int64, int64]

proc initialPrimeGeneratorState(): PrimeGeneratorState =
  result = new(PrimeGeneratorState)
  result.n = 3'i64
  result.last = 2'i64
  result.sieve = newTable[int64,int64]()

proc next(self: PrimeGeneratorState): int64 =

  var prime = self.sieve.getOrDefault(self.n)
  while prime != 0:
    # don't need this key, remove it to save space
    self.sieve.del(self.n)

    # add composite
    var composite = self.n + prime + prime
    while self.sieve.contains(composite):
      composite = composite + prime + prime

    self.sieve[composite] = prime
    self.n = self.n + 2'i64
    prime = self.sieve.getOrDefault(self.n)

  # add composite in prep for next round
  self.sieve[self.n * self.n] = self.n
  result = self.last
  self.last = self.n
  self.n = self.n + 2'i64


type
  FactorFinder = ref object
    known: TableRef[int64, HashSet[int64]]
    nextPrimes: PrimeGeneratorState
    knownPrimes: seq[int64]

proc initialFactorFinder(): FactorFinder =
  result = new(FactorFinder)
  result.nextPrimes = initialPrimeGeneratorState()
  result.knownPrimes = newSeq[int64]()
  var one = initSet[int64]()
  one.incl(1'i64)
  result.known = newTable({1'i64:one})

proc getPrimeFactors(self: FactorFinder, ofn: int64): seq[int64] =
  var factors = newSeq[int64]()
  var quotient = ofn

  for prime in self.knownPrimes:
    if prime > quotient:
      return factors

    var remainder = quotient mod prime
    while remainder == 0'i64:
      quotient = quotient div prime
      remainder = quotient mod prime
      factors.add(prime)

  while true:
    let prime = self.nextPrimes.next()
    self.knownPrimes.add(prime)
    if prime > quotient:
      return factors

    var remainder = quotient mod prime
    while remainder == 0'i64:
      quotient = quotient div prime
      remainder = quotient mod prime
      factors.add(prime)


proc getFactors(self: FactorFinder, ofn: int64): HashSet[int64] =

  let existing = self.known.getOrDefault(ofn)
  if len(existing) > 0:
    return existing

  let pfs = self.getPrimeFactors(ofn)
  var factorSet = initSet[int64]()
  factorSet.incl(1'i64)
  factorSet.incl(ofn)

  for prime in pfs:
    let factor = ofn div prime
    for subfactor in self.getFactors(factor):
      factorSet.incl(subfactor)

  self.known[ofn] = factorSet
  factorSet

proc solve(): int64 =
  let finder = initialFactorFinder()

  var adder = 0'i64
  var tn = 0'i64

  while true:
    adder = adder + 1
    tn = tn + adder

    let tnFactors = finder.getFactors(tn)
    if len(tnFactors) > 1000:
      return tn

echo solve()

File: euler/benchmarks/problem12/cpp/main.cpp
#include <cstdint>
#include <iostream>
#include "robin_hood.h"
#include <vector>

using namespace robin_hood;

namespace factorbench {

class PrimeGenerator {
private:
    int64_t n = 3;
    int64_t last = 2;
    unordered_map<int64_t, int64_t> sieve;
public:
    PrimeGenerator() {}

    int64_t next() {
        while (sieve.find(n) != sieve.end()) {

            auto prime = sieve[n];
            sieve.erase(n);

            int64_t composite = n + prime + prime;
            while (sieve.find(composite) != sieve.end()) {
                composite += prime + prime;
            }
            sieve[composite] = prime;
            n += 2;
        }

        sieve[n * n] = n;
        int64_t r = last;
        last = n;
        n += 2;
        return r;
    }
};


class FactorFinder {
private:
    unordered_map<int64_t, unordered_set<int64_t>> known;
    PrimeGenerator next_primes;
    std::vector<int64_t> known_primes;
public:

    FactorFinder() {
        unordered_set<int64_t> initSet;

        initSet.insert(1);

        known[1] = initSet;
    }

    std::vector<int64_t> get_prime_factors(int64_t of) {

        std::vector<int64_t> factors;
        auto quotient = of;

        for (auto const& prime : known_primes) {
            if (prime > quotient) {
                return factors;
            }

            auto remainder = quotient % prime;
            while (remainder == 0) {
                quotient = quotient / prime;
                remainder = quotient % prime;
                factors.push_back(prime);
            }
        }

        for (;;) {
            auto prime = next_primes.next();
            known_primes.push_back(prime);

            if (prime > quotient) {
                return factors;
            }

            auto remainder = quotient % prime;
            while (remainder == 0) {
                quotient = quotient / prime;
                remainder = quotient % prime;
                factors.push_back(prime);
            }
        }

        return factors;
    }

    unordered_set<int64_t> get_factors(int64_t of) {

        if (known.find(of) != known.end()) {
            return known[of];
        }

        unordered_set<int64_t> factor_set;
        factor_set.insert(1);
        factor_set.insert(of);

        auto pfactors = get_prime_factors(of);
        for (auto prime : pfactors) {

            auto factor = of / prime;
            for (auto subfactor : get_factors(factor)) {
                factor_set.insert(subfactor);
            }

        }

        known[of] = factor_set;
        return factor_set;
    }
};


int64_t solve() {

    FactorFinder pf;

    int64_t adder = 0;
    int64_t tn = 0;

    for (;;) {
        adder += 1;
        tn += adder;

        auto factors = pf.get_factors(tn);
        if (factors.size() > 1000) {
            return tn;
        }
    }
}

} // namespace factorbench

int main() {
    auto answer = factorbench::solve();
    std::cout << "Answer: " << answer << '\n';
    return 0;
}

